### PR TITLE
Enhance default detection and enforce type name deduplication

### PR DESCRIPTION
**Pull request by pdanpdan**
- process default on all structure types
- make sure there are no two structures with the same name

### CODE REVIEW

**Comment by thedumbterminal** [comment]
Thanks for the submission, could you move any listing changes to a separate PR so we can differentiate from the new feature addition and the proposed style change.

File: test/samples/array/expected.json [hunk 1]
```diff
@@ -3,11 +3,11 @@
 	"fields": [
 		{
 			"doc": "example array",
-			"name": "list",
+			"name": "list_required",
```
Comment (thedumbterminal) [inline test/samples/array/expected.json:6]: Could we have a new test sample for the required check. I find smaller examples easier to work with when debugging problems in this codebase.

File: test/samples/complex/expected.json [hunk 2]
```diff
@@ -6,10 +6,10 @@
 	"fields": [
 		{
 			"doc": "the address",
-			"name": "address",
+			"name": "address_required",
```
Comment (thedumbterminal) [inline test/samples/complex/expected.json:9]: See comment above

File: test/samples/enum/expected.json [hunk 3]
```diff
@@ -6,15 +6,31 @@
 	"fields": [
 		{
 			"doc": "the gender",
-			"name": "gender",
+			"name": "gender_required",
```
Comment (thedumbterminal) [inline test/samples/enum/expected.json:9]: Can this be a new sample too

File: test/samples/required/expected.json [hunk 4]
```diff
@@ -25,9 +25,33 @@
...
 					}
 				],
-				"name": "dates_record",
+				"name": "dates_required_record",
 				"type": "record"
 			}
+		},
+		{
+			"doc": "",
+			"name": "dates_optional",
```
Comment (thedumbterminal) [inline test/samples/required/expected.json:34]: Can this be a new sample has having an optional check in a required test is not very readable.

File: src/index.js [hunk 5]
```diff
@@ -71,91 +72,125 @@ jsonSchemaAvro._hasEnum = (schema) => {
 
 jsonSchemaAvro._isRequired = (list, item) => list.includes(item)
 
-jsonSchemaAvro._convertProperties = (schema = {}, required = []) => {
+jsonSchemaAvro._getUniqueName = (name, type, namesCache = {}) => {
```
Comment (thedumbterminal) [inline src/index.js:75]: Could we prefix all names with their parents names to ensure they are unique. `somefield1`, `somefield2` etc. does not relate well to the input as it assumes ordering

**Comment by pdanpdan** [comment]
Hello.
There are no style changes, I just had to update the tests because the the existing ones had no required for the structures, so that the tests were failing now.
Can you please give an example of what you were referring to - I'll infer the required changes from there.

Ah, I think I saw the spacing - I'll take care of them, sorry.

**Comment by thedumbterminal** [comment]
I haven't considered the need to have unique names before, I wonder if the validator i'm using to check this code throws errors when it encounters them. Thanks for raising this.

**Comment by pdanpdan** [comment]
Very good point on prefixing by path. I'll make the changes and come back.
I'll update the existing tests (the input) to have required structures so we can keep the expected and add new tests for optional?
Does this sound good to you?

**Pull request closed without merging**